### PR TITLE
Deprecate LDAP/AD realm settings missing bind password for bind DN

### LINKED ISSUE #47191: Missing `bind_password` or `secure_bind_password` should cause node to fail
When using an LDAP or Active Directory realm, one can set `bind_dn` and the relevant password so that operations against LDAP ( user search, group search etc.) are performed by this service account and not by the authenticated user. 
When one sets `bind_dn` but does not set `bind_password` or `secure_bind_password` , there is no way that the realm configuration will be functional since simple bind operations are not permitted without a password and the underlying LDAP library we use understandably throws an error.

We should fail to start the node, so that our behavior is similar to what we do for other realm misconfigurations.

All the relevant settings are defined in the `PoolingSessionFactorySettings` class. In the [constructor of the `PoolingSessionFactory`](https://github.com/elastic/elasticsearch/blob/12ad399c488f0cc60e19b5e1b29c6d569cb4351a/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ldap/PoolingSessionFactory.java#L60) we read the settings value. We can decide to throw if `bindDn` is not null both both `LEGACY_BIND_PASSWORD` and `SECURE_BIND_PASSWORD` are null

Note: When dynamic realm configuration is in place, we can revisit the validation and which configuration fails the node, but this aims to make the behavior uniform across our realms.

### PR DESCRIPTION
**Pull request by n1v0lg**
LDAP/AD authentication realms can be configured to authenticate through
LDAP via a bind user. For this it's necessary to set a bind DN (via
bind_dn) together with a bind password (via bind_password or
secure_bind_password). Setting a bind DN without a bind password will
cause all LDAP/AD realm authentication to fail, leaving the node
non-operational. This PR introduces a deprecation warning when bind_dn
is set without a bind password.

Closes #47191

### CODE REVIEW

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ldap/PoolingSessionFactory.java [hunk 1]
```diff
@@ -79,9 +80,8 @@
                         + RealmSettings.getFullSettingKey(config, SECURE_BIND_PASSWORD)
                         + "]"
                 );
-            } else {
-                bindPassword = CharArrays.toUtf8Bytes(config.getSetting(LEGACY_BIND_PASSWORD).getChars());
             }
+            bindPassword = CharArrays.toUtf8Bytes(config.getSetting(LEGACY_BIND_PASSWORD).getChars());
```
Comment (n1v0lg) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ldap/PoolingSessionFactory.java:84]: Small readability tweak: moved this out of the `else` block because the corresponding `if` block always throws. With this tweak we avoid a level of nesting

File: docs/changelog/85326.yaml [hunk 2]
```diff
@@ -0,0 +1,6 @@
+pr: 85326
+summary: "Fix: fail node start if LDAP/AD realm has bind DN set but is missing bind password"
```
Comment (tvernum) [inline docs/changelog/85326.yaml:2]: The `Fix:` prefix is unnecessary.
The PR is labelled as `>bug`, so the changelog file already knows that this is `type: bug` and the release notes will include it in the relevant section.
```suggestion
summary: "Fail node start if LDAP/AD realm has bind DN set but is missing bind password"
```

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ldap/PoolingSessionFactory.java [hunk 3]
```diff
@@ -91,6 +91,17 @@
...
             bindCredentials = new SimpleBindRequest();
         } else {
+            if (bindPassword == null) {
+                throw new IllegalArgumentException(
+                    "When ["
+                        + RealmSettings.getFullSettingKey(config, BIND_DN)
+                        + "] is set you must also specify ["
+                        + RealmSettings.getFullSettingKey(config, LEGACY_BIND_PASSWORD)
+                        + "] or ["
+                        + RealmSettings.getFullSettingKey(config, SECURE_BIND_PASSWORD)
```
Comment (tvernum) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ldap/PoolingSessionFactory.java:101]: We should list `SECURE_BIND_PASSWORD` before `LEGACY_BIND_PASSWORD` as it is the preferred (non deprecated) setting and people are more likely to follow the first suggested fix.

**Review by tvernum — COMMENTED** [review]
LGTM, but I want us to have a discussion about whether we think a change like this is breaking (since it moves a runtime error to a startup error) before we commit it.

**Comment by n1v0lg** [comment]
@tvernum  Thanks for raising this! 

Strictly speaking this is a breaking change since nodes that currently boot will not boot anymore if this PR is merged. As such it's at least worth highlighting this new behavior to users in the release notes.

Alternatively, we could always go with logging an error instead of failing boot. Two downsides here: (1) the misconfiguration is harder to spot, and (2) this is inconsistent with how we handle other critical realm misconfigurations: for LDAP and other realms, we fail boot in such cases (e.g., when both legacy and secure bind passwords are set). 

I'm wondering to what extent this breaking change would impact users in practice: if users configured their realm correctly, there is no impact. Furthermore, it seems unlikely that users are actively relying on misconfigured nodes running in their cluster. If LDAP is the primary realm, virtually all actions against the node will fail which is functionally similar to a failed boot, just harder to debug. One edge case I see is if the node has other realms configured: in that case, the node would still be partially operational before this PR (since actions that auth against the other realms would still work), but not so afterwards. Even in this scenario though, the node is still misconfigured and likely to be fixed soon after boot. 

TL;DR this is a breaking change with limited scope (only applies to already broken configurations) and an easy resolution. My intuition is that it's worthwhile. However, I'm onboard with a backwards compatible error log since the breaking change is technically unnecessary.

Let me know what you think!

**Comment by n1v0lg** [comment]
Summary on discussion during team meeting:

- We want to start with an error log to avoid breaking change
- Would ideally make this error log a deprecation warning
- Longer term, we want to make the breaking change and fail to boot the node

File: docs/changelog/85326.yaml [hunk 4]
```diff
@@ -0,0 +1,6 @@
+pr: 85326
+summary: "Log error if LDAP/AD realm has bind DN set but is missing bind password"
+area: Authentication
+type: bug
```
Comment (n1v0lg) [inline docs/changelog/85326.yaml:4]: I'm not sure what our convention is here: is this still a bug fix, even if it's just a log line, or should we rebrand as enhancement?

**Review by tvernum — APPROVED** [review]
LGTM

**Comment by n1v0lg** [comment]
@elasticmachine update branch plz

**Comment by n1v0lg** [comment]
@elasticmachine update branch bitte

File: docs/changelog/85326.yaml [hunk 5]
```diff
@@ -0,0 +1,24 @@
+pr: 85326
+summary: Deprecate LDAP/AD realm settings missing bind password for bind DN
+area: Authentication
+type: deprecation
+issues:
+ - 47191
+deprecation:
+  title: Deprecate LDAP/AD realm settings missing bind password for bind DN
+  area: Cluster and node setting
```
Comment (n1v0lg) [inline docs/changelog/85326.yaml:9]: `Authentication` and `Security` are among the possible values here, it turns out. Possibly values are here: https://github.com/elastic/elasticsearch/blob/master/build-tools-internal/src/main/resources/changelog-schema.json#L178

File: docs/changelog/85326.yaml [hunk 6]
```diff
@@ -0,0 +1,24 @@
+pr: 85326
+summary: Deprecate LDAP/AD realm settings missing bind password for bind DN
+area: Authentication
+type: deprecation
+issues:
+ - 47191
+deprecation:
```
Comment (n1v0lg) [inline docs/changelog/85326.yaml:7]: My current understand is that this won't automatically create migration docs and will stay in the release notes instead. Please let me know if that's not the case.

**Comment by n1v0lg** [comment]
@elasticmachine update branch

**Comment by n1v0lg** [comment]
@elasticmachine update branch

**Review by tvernum — APPROVED** [review]
LGTM

**Comment by n1v0lg** [comment]
@elasticmachine run elasticsearch-ci/part-2-fips por favor

**Comment by n1v0lg** [comment]
@elasticmachine run elasticsearch-ci/part-2-fips again por favor

**Comment by n1v0lg** [comment]
@elasticmachine update branch

**Review by tvernum — APPROVED** [review]
LGTM

**Comment by n1v0lg** [comment]
@elasticmachine update branch plz

**Comment by n1v0lg** [comment]
@elasticmachine update branch
